include/swift/ABI/Task.h

```diff
@@ -293,12 +663,19 @@ class AsyncTask : public HeapObject, public Job {
 
   FutureFragment *futureFragment() {
     assert(isFuture());
+
+    auto offset = reinterpret_cast<uintptr_t>(this); // TODO: char* instead?
+    offset += sizeof(AsyncTask);
+
     if (hasChildFragment()) {
-      return reinterpret_cast<FutureFragment *>(
-          reinterpret_cast<ChildFragment*>(this + 1) + 1);
+      offset += sizeof(ChildFragment);
     }
 
-    return reinterpret_cast<FutureFragment *>(this + 1);
+    if (isTaskGroup()) {
+      offset += sizeof(GroupFragment);
+    }
+
+    return reinterpret_cast<FutureFragment *>(offset);
```

Comment:
@DougGregor I think I got this right; last time we talked this was an intense if/else dance if you remember.